cmd/createorupdate/update.go

```diff
@@ -0,0 +1,231 @@
+// extra VM.
+func updatePlusOne(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin, ssc compute.VirtualMachineScaleSetsClient, vmc compute.VirtualMachineScaleSetVMsClient, role api.AgentPoolProfileRole) error {
+	count := getCount(cs, role)
+
+	// store a list of all the VM instances now, so that if we end up creating
+	// new ones (in the crash recovery case, we might not), we can detect which
+	// they are
+	vms, err := listVMs(ctx, cs, vmc, role)
+	if err != nil {
+		return err
+	}
+
+	vmsBefore := map[string]struct{}{}
+	for _, vm := range vms {
+		vmsBefore[*vm.InstanceID] = struct{}{}
+	}
+
+	log.Infof("setting ss-%s capacity to %d", role, count+1)
+	future, err := ssc.Update(ctx, cs.Properties.AzProfile.ResourceGroup, "ss-"+string(role), compute.VirtualMachineScaleSetUpdate{
+		Sku: &compute.Sku{
+			Capacity: to.Int64Ptr(int64(count) + 1),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	err = future.WaitForCompletion(ctx, ssc.Client)
+	if err != nil {
+		return err
+	}
+
+	vms, err = listVMs(ctx, cs, vmc, role)
+	if err != nil {
+		return err
+	}
+
+	// wait for newly created VMs to reach readiness (n.b. one alternative to
+	// this approach would be for the CSE to not return until the node is
+	// ready, but that is also problematic)
+	for _, vm := range vms {
+		if _, found := vmsBefore[*vm.InstanceID]; !found {
+			log.Infof("waiting for %s to be ready", *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+			err = p.WaitForReady(ctx, cs, role, *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	err = updateInPlace(ctx, cs, p, ssc, vmc, role)
+	if err != nil {
+		return err
+	}
+
+	// remove surplus VMs
+	for _, vm := range vms[count:] {
+		log.Infof("draining %s", *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+		err = p.Drain(ctx, cs, role, *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+		if err != nil {
+			return err
+		}
+
+		log.Infof("deleting %s", *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+		future, err := vmc.Delete(ctx, cs.Properties.AzProfile.ResourceGroup, "ss-"+string(role), *vm.InstanceID)
+		if err != nil {
+			return err
+		}
+
+		err = future.WaitForCompletion(ctx, vmc.Client)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// updateInPlace updates one by one all the VMs of a scale set, in place.
+func updateInPlace(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin, ssc compute.VirtualMachineScaleSetsClient, vmc compute.VirtualMachineScaleSetVMsClient, role api.AgentPoolProfileRole) error {
+	vms, err := listVMs(ctx, cs, vmc, role)
+	if err != nil {
+		return err
+	}
+
+	for _, vm := range vms {
+		log.Infof("draining %s", *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+		err = p.Drain(ctx, cs, role, *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+		if err != nil {
+			return err
+		}
+
+		{
```

Comment:
@kargakis these are really good questions. @mjudeikis, @pweil- fyi:

I think that Reimage and UpdateInstances return the VM to its power state before the API call.  I think if you do them back-to-back with the VM running you end up booting twice which is even slower than what we're doing now, and might cause a cluster blip if the node manages to register itself in the interim.  Hence wrapping with Deallocate/Start.

UpdateInstances does not necessarily restart the VM - it can just rerun a new CSE.  I think the error cases and drift are much harder to handle in that case, but I'm open to discussion.  Hence I liked the idea (at least to begin with) of Deallocate, UpdateInstances, Reimage, Start.

It is horribly slow, however, but I think that can be helped a lot with the bitmap idea we were discussing this morning.

As the implementation stands, it effectively makes "update" a safe cluster reset operation, and I quite like that.

Reply:
Yes, although AFAICS it seems to run quite quickly in that case.